apps/prairielearn/assets/scripts/workspaceClient.ts

```diff
@@ -29,7 +30,12 @@ $(function () {
     30 * 60,
   );
 
-  const socket = io('/workspace');
+  const socket = io('/workspace', {
+    auth: {
+      token: socketToken,
+      workspace_id: workspaceId,
+    },
+  });
```

Comment:
I'm experimenting with a new auth flow here. On other pages, we include information about the resource and a token in an initial `join*` message. However, by including this information in the socket handshake on this `auth` object, we can simplify things a bit. If a client connects, has a valid `*_id` and a corresponding token, we can immediately connect them and emit any necessary state. How does this sound?

If we like this, I may go update our other pages (server jobs and external grading results) to use the same method for consistency.

Reply:
That said: we do actually have separate websocket "rooms" for different server jobs _within_ a given sequence, and we use multiple of them on that page. So this might not generalize well. Though maybe we can rework that so that we send all messages (for both the job sequence and constituent jobs) over a single room?

Reply:
Good find! Actually, based on my reading of that, we're already using multiplexing since `multiplex` defaults to `true` and we don't pass `forceNew: true`.

Reply:
I don't think we're actually using multiplexing because two objects on the same namespace (`/workspace` above) [create separate managers](https://socket.io/docs/v4/client-options/#forcenew).

Reply:
Ah, you're right, I see that described here as well: https://socket.io/docs/v4/namespaces/#client-initialization. It seems like it isn't possible to ever have two sockets on the same namespace share a manager/connection.

Reply:
This could be done with a namespace that includes the workspace ID...

Reply:
https://socket.io/docs/v4/namespaces/#dynamic-namespaces

Reply:
Ahh, that's an interesting idea! Might be cleaner than the rooms stuff we've been doing...